### PR TITLE
Fix: DISTINCT in order with zero values in non-sorted columns

### LINKED ISSUE #49622: Unexpected "DISTINCT" result with empty string
+ ClickHouse Version

    `23.4.2.11`

+ DDL
    ```sql
    CREATE TABLE test
    (
        c1 String,
        c2 String,
        c3 String
    )
    ENGINE = ReplacingMergeTree
    ORDER BY (c1, c3);
    
    INSERT INTO test(c1, c2, c3) VALUES
        ('', '', '1'), ('', '', '2'),('v1', 'v2', '3'),('v1', 'v2', '4'),('v1', 'v2', '5');
    ```

```sql

SELECT c1, c2, c3 FROM test GROUP BY c1, c2, c3;
┌─c1─┬─c2─┬─c3─┐
│    │    │ 2  │
│ v1 │ v2 │ 4  │
│ v1 │ v2 │ 3  │
│    │    │ 1  │
│ v1 │ v2 │ 5  │
└────┴────┴────┘

SELECT DISTINCT c1, c2, c3 FROM test;
┌─c1─┬─c2─┬─c3─┐
│    │    │ 1  │
│ v1 │ v2 │ 3  │
│ v1 │ v2 │ 4  │
│ v1 │ v2 │ 5  │
└────┴────┴────┘

```
Why does the query result of this DISTINCT clause not contain rows where c3 is 2

### PR DESCRIPTION
**Pull request by devcrafter**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixes bug in DISTINCT in order with zero values in non-sorted columns. Fixes #49622

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit 68fcc39292ed6d515dba3af4ad25148a141ba3b7 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49636/68fcc39292ed6d515dba3af4ad25148a141ba3b7/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Common/HashTable/ClearableHashSet.h [hunk 1]
```diff
@@ -88,6 +88,9 @@ class ClearableHashSet
     {
         ++this->version;
         this->m_size = 0;
+
+        /// clear ZeroValueStorage
+        this->clearHasZero();
```
Comment (alexey-milovidov) [inline src/Common/HashTable/ClearableHashSet.h:93]: Should we check for hasZero?

**Review by alexey-milovidov — APPROVED** [review]

**Comment by devcrafter** [comment]
AST fuzzer (msan) - created https://github.com/ClickHouse/ClickHouse/issues/49672

**Comment by alexey-milovidov** [comment]
@devcrafter there is still one unresolved question: it's unclear why zero-value storage is needed for ClearableHashMap.

**Pull request merged**